File: 05/c/main.c
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdbool.h>

#define SWAP(x, y) { int SWAP = x; x = y; y = SWAP; };

#define N 1000
int board[N][N] = {0};

struct line {
  int x1;
  int y1;
  int x2;
  int y2;
};

#define lines_cap 512
struct line lines[lines_cap];
int lines_size = 0;

int main(int argc, char *argv[]) {
  assert(argc == 2 && "Must provide one argument => input.txt path");

  printf("Input file: %s\n\n", argv[1]);
  FILE *file = fopen(argv[1], "r");

  int x1 = 0, y1 = 0, x2 = 0, y2 = 0;
  while (fscanf(file, "%d,%d -> %d,%d", &x1, &y1, &x2, &y2) == 4) {
    lines[lines_size].x1 = x1;
    lines[lines_size].x2 = x2;
    lines[lines_size].y1 = y1;
    lines[lines_size].y2 = y2;
    lines_size++;
  }

  // Problema 1
  {
    int count = 0;

    int curr = 0;
    while (curr < lines_size) {
      int x1 = lines[curr].x1;
      int x2 = lines[curr].x2;
      int y1 = lines[curr].y1;
      int y2 = lines[curr].y2;
      curr++;

      if (x1 != x2 && y1 != y2) continue; // diagonal

      if (x1 > x2) SWAP(x1, x2)
      if (y1 > y2) SWAP(y1, y2)

      for (int x = x1; x <= x2; x++) {
        for (int y = y1; y <= y2; y++) {
          if (board[y][x] == 1) count++;
          board[y][x]++;
        }
      }
    }

    printf("Answer 1: %d\n", count);
  }

  memset(board, 0, N * N * sizeof(int));

  // Problema 2
  {
    int count = 0;

    int curr = 0;
    while (curr < lines_size) {
      int x1 = lines[curr].x1;
      int x2 = lines[curr].x2;
      int y1 = lines[curr].y1;
      int y2 = lines[curr].y2;
      curr++;

      if (x1 != x2 && y1 != y2) { // diagonal
        int yDir = y2 > y1 ? 1 : -1;

        if (x2 > x1) {
          for (int x = x1; x <= x2; x++) {
            int y = y1 + (x - x1) * yDir;
            if (board[y][x] == 1) count++;
            board[y][x]++;
          }
        } else {
          for (int x = x1; x >= x2; x--) {
            int y = y1 + (x1 - x) * yDir;
            if (board[y][x] == 1) count++;
            board[y][x]++;
          }
        }
      } else {
        if (x1 > x2) SWAP(x1, x2)
        if (y1 > y2) SWAP(y1, y2)

        for (int x = x1; x <= x2; x++) {
          for (int y = y1; y <= y2; y++) {
            if (board[y][x] == 1) count++;
            board[y][x]++;
          }
        }
      }
    }

    printf("Answer 2: %d\n", count);
  }

  fclose(file);

  return 0;
}

File: 05/js/main.js
const fs = require('fs')
const {join} = require('path')

const file = fs.readFileSync(join(__dirname, '../input.txt'), 'utf8')
const inputs = file.split('\n').slice(0, -1).map(line => line.replace(' -> ', ',').split(',').map(n => Number(n)))
const board = []

const size = 1000
for (let x = 0; x < size; x++) {
  board[x] = []
  for (let y = 0; y < size; y++) {
    board[x][y] = 0
  }
}

// Problem 1
{
  const inputs1 = [...inputs]
  let count = 0

  while (inputs1.length) {
    let [x1, y1, x2, y2] = inputs1.shift()

    if (x1 != x2 && y1 != y2) continue // diagonal

    if (x1 > x2) [x1, x2] = [x2, x1]
    if (y1 > y2) [y1, y2] = [y2, y1]

    for (let x = x1; x <= x2; x++) {
      for (let y = y1; y <= y2; y++) {
        if (board[y][x] == 1) count++
        board[y][x]++
      }
    }
  }

  console.log(`Answer 1: ${count}`);
}

// Clear board
for (row of board) {
  for (let i = 0; i < row.length; i++) {
    row[i] = 0
  }
}

// Problem 2
{
  let count = 0

  while (inputs.length) {
    let [x1, y1, x2, y2] = inputs.shift()


    if (x1 != x2 && y1 != y2) { // diagonal
      let yDir = y2 > y1 ? 1 : -1

      if (x2 > x1) {
        for (let x = x1; x <= x2; x++) {
          const y = y1 + (x - x1) * yDir
          if (board[y][x] == 1) count++
          board[y][x]++
        }
      } else {
        for (let x = x1; x >= x2; x--) {
          const y = y1 + (x1 - x) * yDir
          if (board[y][x] == 1) count++
          board[y][x]++
        }
      }
    } else {
      if (x1 > x2) [x1, x2] = [x2, x1]
      if (y1 > y2) [y1, y2] = [y2, y1]

      for (let x = x1; x <= x2; x++) {
        for (let y = y1; y <= y2; y++) {
          if (board[y][x] == 1) count++
          board[y][x]++
        }
      }
    }
  }

  console.log(`Answer 2: ${count}`);
}
